walkFiles consistently relative or absolute

jean-philippe-martin opened the pull request:
The expected behavior from Files.walkFileTree is that if the starting path is absolute, then all the visited paths are absolute. If the starting path is relative, then all the visited paths are relative.

This PR adds a test for this behavior, and shows that our current code
fails for an absolute path: it returns a mix of relative and absolute
paths.

This PR also adds a code change to fix the problem, so the actual
behavior matches what is expected.

This should fix issue #3772

pongad commented on google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/it/ITGcsNio.java:
```diff
@@ -620,6 +667,32 @@ private String randomSuffix() {
     return "-" + rnd.nextInt(99999);
   }
 
+  private static class postTraversalWalker extends SimpleFileVisitor<Path> {
```
This is a class so `PostTraversalWalker`?

jean-philippe-martin replied:
You're right, I fixed it now.

pongad commented on google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/it/ITGcsNio.java:
```diff
@@ -538,6 +539,52 @@ public void testListFiles() throws IOException {
+      for (Path p : relativePaths) {
+        assertWithMessage("Should have been relative: " + p.toString()).that(p.isAbsolute()).isFalse();
+      }
+      assertThat(relativePaths.size()).isEqualTo(5);
```
Do you think it's worth explicitly listing out the paths?

jean-philippe-martin replied:
I think it's preferable to keep this test focused on only one thing. There'a already another test that checks that walkFiles finds all the files.

olifly commented:
Hi.

I would think that these modifications, along with the ones in https://github.com/googleapis/google-cloud-java/pull/3775, would allow the following unit test to pass:

```java
  @Test
  public void testWalkFileTree() throws IOException {
    Path baseDir = Paths.get(URI.create("gs://bucket/liverpool"));
    //Base directory with an ending forwards slash is not passing either
    //Path baseDir = Paths.get(URI.create("gs://bucket/liverpool/"));
    Path file1 = baseDir.resolve("firmino.player");
    Path file2 = baseDir.resolve("salah.player");
    Files.createFile(file1);
    Files.createFile(file2);

    Set<Path> expectedDirs = Sets.newHashSet(baseDir);
    Set<Path> expectedFiles = Sets.newHashSet(file1,file2);
    final Set<Path> foundDirs = Sets.newHashSet();
    final Set<Path> foundFiles = Sets.newHashSet();

    FileVisitor<Path> testVisitor = new SimpleFileVisitor<Path>()  {

      @Override
      public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
        assertThat(foundDirs).doesNotContain(dir);
        foundDirs.add(dir);
        return FileVisitResult.CONTINUE;
      }

      @Override
      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
        foundFiles.add(file);
        return FileVisitResult.CONTINUE;
      }

      @Override
      public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
        assertThat(foundDirs).contains(dir);
        return FileVisitResult.CONTINUE;
      }
    };

    Files.walkFileTree(baseDir,Sets.newHashSet(FileVisitOption.FOLLOW_LINKS),Integer.MAX_VALUE,testVisitor);

    assertThat(foundFiles).containsExactlyElementsIn(expectedFiles);
    assertThat(foundDirs).containsExactlyElementsIn(expectedDirs);

  }
```
Files.walkFileTree visits the Basedir twice, first without an ending slash, and then with an ending slash.

Best regards,
Ólafur Haukur Flygenring

jean-philippe-martin commented:
@olifly I see there is a bug in this code snippet: `newHashSet` expects an iterator, not a single element. Instead of a set that contains the path, you get a set that contains a list of the elements that compose the path. So for example /foo/bar/baz/ will get you a set with foo, bar, and baz.

Once I fixed the code snippet, it passed with just the code in this PR.

olifly commented:
> @olifly I see there is a bug in this code snippet: `newHashSet` expects an iterator, not a single element. Instead of a set that contains the path, you get a set that contains a list of the elements that compose the path. So for example /foo/bar/baz/ will get you a set with foo, bar, and baz.
> 
> Once I fixed the code snippet, it passed with just the code in this PR.

Hi.

Are you sure it's passing?  Changing the test to insert the expected Paths one at a time into the Sets is not making it pass for me.  Files.walkFileTree is still calling preVisitDirectory twice for the base dir, once without a forward slash and once with it.  

Here's my modified code

```java
@Test
  public void testWalkFileTree() throws IOException {
    Path baseDir = Paths.get(URI.create("gs://bucket/liverpool"));
    Path file1 = baseDir.resolve("firmino.player");
    Path file2 = baseDir.resolve("salah.player");
    Files.createFile(file1);
    Files.createFile(file2);

    Set<Path> expectedDirs = Sets.newHashSet();
    expectedDirs.add(baseDir);
    Set<Path> expectedFiles = Sets.newHashSet();
    expectedFiles.add(file1);
    expectedFiles.add(file2);

    final Set<Path> foundDirs = Sets.newHashSet();
    final Set<Path> foundFiles = Sets.newHashSet();

    FileVisitor<Path> testVisitor = new SimpleFileVisitor<Path>()  {

      @Override
      public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
        assertThat(foundDirs).doesNotContain(dir);
        foundDirs.add(dir);
        return FileVisitResult.CONTINUE;
      }

      @Override
      public FileVisitResult visitFile(Path file, BasicFileAttributes attrs) {
        foundFiles.add(file);
        return FileVisitResult.CONTINUE;
      }

      @Override
      public FileVisitResult postVisitDirectory(Path dir, IOException exc) {
        assertThat(foundDirs).contains(dir);
        return FileVisitResult.CONTINUE;
      }
    };

    Files.walkFileTree(baseDir,Sets.newHashSet(FileVisitOption.FOLLOW_LINKS),Integer.MAX_VALUE,testVisitor);

    assertThat(foundFiles).containsExactlyElementsIn(expectedFiles);
    assertThat(foundDirs).containsExactlyElementsIn(expectedDirs);

  }
```

Best Regards,
Ólafur Haukur Flygenring

jean-philippe-martin commented:
@olifly you're right, the "/liverpool" case (where we have to merge the two PRs) still didn't work right. I updated #3775 to fixt listing of a pretend-folder that doesn't end in slash (and added a test case for it). Thank you very much! Once you add that change, I expect your code should work.

jean-philippe-martin commented on google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/it/ITGcsNio.java:
```diff
@@ -538,6 +539,52 @@ public void testListFiles() throws IOException {
+      paths.add(fs.getPath("dir/dir2/another_angel"));
+      paths.add(fs.getPath("atroot"));
+      paths.addAll(goodPaths);
+      goodPaths.add(fs.getPath("dir/dir2/"));
```
You're correct. Removing it.

jean-philippe-martin commented:
@chingor13 the functions returns the folders and files traversed, so the 5 paths are:
`dir/`, `dir/angel`, `dir/alone`, `dir/dir2/`, `dir/dir2/another_angel`.

I added them as comment to the code in case another reader wonders too.

(sorry, for some reason GitHub isn't letting me reply to your comment directly).

olifly commented:
@jean-philippe-martin - The changes you did to pull request #3775 did indeed solve the issue.  Thanks a bunch.  

Can't wait for these changes to be merged upstream :)

Best Regards
Ólafur Haukur Flygenring.

jean-philippe-martin commented:
@olifly thanks for checking, that's great! I'm glad this works for you now. I see the reviewers have already merged in this PR so we're halfway there!

jean-philippe-martin commented:
Thank you @chingor13 !
